js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataThread.tsx

```diff
@@ -0,0 +1,255 @@
+import {
+  AssetGraphLiveQuery,
+  AssetGraphLiveQueryVariables,
+  AssetNodeLiveFragment,
+} from './types/AssetLiveDataThread.types';
+import {BATCHING_INTERVAL} from './util';
+
+export type AssetLiveDataThreadID = 'default' | 'sidebar' | 'asset-graph';
+
+export class AssetLiveDataThread {
+  private isFetching: boolean = false;
+  private listenersCount: {[key: string]: number};
+  private client: ApolloClient<any>;
+  private isLooping: boolean = false;
+  private interval?: ReturnType<typeof setTimeout>;
+  private manager: AssetLiveDataThreadManager;
+  public pollRate: number = 30000;
+
+  protected static _threads: {[key: string]: AssetLiveDataThread} = {};
+
+  constructor(client: ApolloClient<any>, manager: AssetLiveDataThreadManager) {
+    this.client = client;
+    this.listenersCount = {};
+    this.manager = manager;
+  }
+
+  public setPollRate(pollRate: number) {
+    this.pollRate = pollRate;
+  }
+
+  public subscribe(key: string) {
+    this.listenersCount[key] = this.listenersCount[key] || 0;
+    this.listenersCount[key] += 1;
+    this.startFetchLoop();
+  }
+
+  public unsubscribe(key: string) {
+    this.listenersCount[key] -= 1;
+    if (this.listenersCount[key] === 0) {
+      delete this.listenersCount[key];
+    }
+    if (this.getObservedKeys().length === 0) {
+      this.stopFetchLoop();
+    }
+  }
+
+  public getObservedKeys() {
+    return Object.keys(this.listenersCount);
+  }
+
+  public startFetchLoop() {
+    if (this.isLooping) {
+      return;
+    }
+    this.isLooping = true;
+    const fetch = () => {
+      const nextAssets = this.manager.determineAssetsToFetch(this.getObservedKeys());
+      if (nextAssets.length) {
+        this._batchedQueryAssets(nextAssets);
+      }
+    };
+    setTimeout(fetch, BATCHING_INTERVAL);
+    this.interval = setInterval(fetch, 5000);
+  }
+
+  public stopFetchLoop() {
+    if (!this.isLooping) {
+      return;
+    }
+    this.isLooping = false;
+    clearInterval(this.interval);
+    this.interval = undefined;
+  }
+
+  private async queryAssetKeys(assetKeys: AssetKeyInput[]) {
+    const {data} = await this.client.query<AssetGraphLiveQuery, AssetGraphLiveQueryVariables>({
+      query: ASSETS_GRAPH_LIVE_QUERY,
+      fetchPolicy: 'network-only',
+      variables: {
+        assetKeys,
+      },
+    });
+    const nodesByKey: Record<string, AssetNodeLiveFragment> = {};
+    const liveDataByKey: Record<string, LiveDataForNode> = {};
+    data.assetNodes.forEach((assetNode) => {
+      const id = tokenForAssetKey(assetNode.assetKey);
+      nodesByKey[id] = assetNode;
+    });
+    data.assetsLatestInfo.forEach((assetLatestInfo) => {
+      const id = tokenForAssetKey(assetLatestInfo.assetKey);
+      liveDataByKey[id] = buildLiveDataForNode(nodesByKey[id]!, assetLatestInfo);
+    });
+
+    this.manager._updateCache(liveDataByKey);
+    return liveDataByKey;
+  }
+
+  private async _batchedQueryAssets(assetKeys: AssetKeyInput[]) {
```

Comment:
It seems like this function takes a param, but because it completes and calls itself recursively with `this.manager.determineAssetsToFetch(this.getObservedKeys())`, I think it would be more correct to remove the param and make that the first line of the function:

```
assetKeys = this.manager.determineAssetsToFetch(this.getObservedKeys());
```

(Since even if you call it with different asset keys, it doesn't stop with the assetKeys you give it and essentially reverts to using the above after the first call)